pkg/ctl/install/install.go

```diff
@@ -0,0 +1,444 @@
+		return err
+	}
+	cleanCloneDir = true
+
+	logger.Info("Flux will operate properly only once it has SSH access to: %s", fi.opts.templateParams.GitURL)
+	logger.Info("please add the following Flux public SSH key to your repository:\n%s", fluxSSHKey.Key)
+	return nil
+}
+
+func (fi *fluxInstaller) cloneRepo(ctx context.Context) (string, error) {
+	cloneDir, err := ioutil.TempDir(os.TempDir(), "eksctl-install-flux-clone")
+	if err != nil {
+		return "", fmt.Errorf("cannot create temporary directory: %s", err)
+	}
+	cloneCtx, cloneCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer cloneCtxCancel()
+	args := []string{"clone", "-b", fi.opts.templateParams.GitBranch, fi.opts.templateParams.GitURL, cloneDir}
+	err = runGitCmd(cloneCtx, cloneDir, args...)
+	return cloneDir, err
+}
+
+func (fi *fluxInstaller) addCommitAndPushFluxManifests(ctx context.Context, cloneDir string) error {
+	// git add
+	addCtx, addCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer addCtxCancel()
+	if err := runGitCmd(addCtx, cloneDir, "add", "--", fi.opts.gitFluxPath); err != nil {
+		return err
+	}
+
+	// Confirm there is something to commit, otherwise move on
+	diffCtx, diffCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer diffCtxCancel()
+	if err := runGitCmd(diffCtx, cloneDir, "diff", "--cached", "--quiet", "--", fi.opts.gitFluxPath); err == nil {
+		logger.Info("Nothing to commit (the repository contained identical manifests), moving on")
+		return nil
+	} else if _, ok := err.(*exec.ExitError); !ok {
+		return err
+	}
+
+	// git commit
+	commitCtx, commitCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer commitCtxCancel()
+	args := []string{"commit",
+		"-m", "Add Initial Flux configuration",
+		fmt.Sprintf("--author=%s <%s>", fi.opts.templateParams.GitUser, fi.opts.templateParams.GitEmail),
+	}
+	if err := runGitCmd(commitCtx, cloneDir, args...); err != nil {
+		return err
+	}
+
+	// git push
+	pushCtx, pushCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer pushCtxCancel()
+	return runGitCmd(pushCtx, cloneDir, "push")
+}
+
+func (fi *fluxInstaller) applyManifests(manifestsMap map[string][]byte) error {
+	// TODO: initialise the client elsewhere so that a mock client can easily be dependency-injected for testing purposes.
+	client, err := kubernetes.NewRawClient(fi.k8sClientSet, fi.k8sRestConfig)
+	if err != nil {
+		return err
+	}
+
+	// If the namespace needs to be created, do it first before any other
+	// resource which should potentially be created within this namespace.
+	// Otherwise, creation of these resources will fail.
+	if namespace, ok := manifestsMap[namespaceFileName]; ok {
+		if err := createOrReplace(namespace, client); err != nil {
+			return err
+		}
+		delete(manifestsMap, namespaceFileName)
+	}
+	manifests := kubernetes.ConcatManifestValues(manifestsMap)
+	return createOrReplace(manifests, client)
+}
+
+func createOrReplace(manifest []byte, client *kubernetes.RawClient) error {
```

Comment:
Should I move this to some utility package? It feels like something which could be re-used.

Reply:
There is similar code in kubernetes/client.go

Reply:
Hmm, what do you mean? I could see the `RawClient` and `RawResource` `struct`s in `pkg/kubernetes/client.go`, but nothing similar, nor higher-level nor very façade-pattern-ish. Moreover `RawResource#CreateOrReplace` does not seem used very wildly:
```console
$  git grep CreateOrReplace
pkg/addons/default/aws_node.go:         status, err := resource.CreateOrReplace(plan)
pkg/addons/default/aws_node_test.go:                            _, err = rc.CreateOrReplace(false)
pkg/addons/default/coredns.go:          status, err := resource.CreateOrReplace(plan)
pkg/addons/default/coredns_test.go:                             _, err = rc.CreateOrReplace(false)
pkg/ctl/install/install.go:     status, err := resource.CreateOrReplace(false)
pkg/kubernetes/client.go:// CreateOrReplace will check if the given resource exists, and create or update it as needed
pkg/kubernetes/client.go:func (r *RawResource) CreateOrReplace(plan bool) (string, error) {
pkg/kubernetes/client_test.go:                          _, err := rc.CreateOrReplace(false)
pkg/kubernetes/client_test.go:                          _, err := rc.CreateOrReplace(false)
pkg/kubernetes/client_test.go:                          _, err := rc.CreateOrReplace(false)
pkg/kubernetes/client_test.go:                          _, err = rc.CreateOrReplace(false)
pkg/kubernetes/client_test.go:                          _, err = rc.CreateOrReplace(false)
```

Reply:
I've extracted these functions and added a higher level `RawClient#ClientOrUpdate` method which now does all the heavy lifting and can be more easily re-used. See: 6c0f9886